File: backend/src/main/java/com/riskdashboard/service/RiskScoringService.java
package com.riskdashboard.service;

import com.riskdashboard.model.FraudSignalAssessment;
import com.riskdashboard.model.RiskAssessment;
import com.riskdashboard.model.RiskTier;
import com.riskdashboard.model.TransactionData.Transaction;

import java.time.Instant;
import java.util.Comparator;
import java.util.List;

/**
 * Orchestrates the four fraud signal calculators to produce a full {@link RiskAssessment}
 * for a given account.
 *
 * <p>Scoring is deterministic: the same account transactions and reference time always
 * produce an identical score, tier, and ordered signal list.</p>
 *
 * <p>The {@code totalRiskScore} is the sum of all four weighted contributions, clamped
 * to {@code [0.0, 1.0]}. Signals in the returned assessment are ordered by
 * {@code weightedContribution} descending.</p>
 */
public class RiskScoringService {

    private final FraudSignalCalculatorService signalCalculator;
    private final RiskScoringConfig config;

    public RiskScoringService(FraudSignalCalculatorService signalCalculator,
                               RiskScoringConfig config) {
        if (signalCalculator == null) {
            throw new IllegalArgumentException("FraudSignalCalculatorService must not be null");
        }
        if (config == null) {
            throw new IllegalArgumentException("RiskScoringConfig must not be null");
        }
        this.signalCalculator = signalCalculator;
        this.config = config;
    }

    /**
     * Computes a full risk assessment for the given account at the given reference time.
     *
     * @param accountId     account identifier; must not be null or blank
     * @param transactions  the account's transaction list; must not be null
     * @param referenceTime the instant representing "now"; must not be null
     * @return a complete, deterministic {@link RiskAssessment}
     * @throws IllegalArgumentException if any argument is null or {@code accountId} is blank
     */
    public RiskAssessment assess(String accountId,
                                  List<Transaction> transactions,
                                  Instant referenceTime) {
        if (accountId == null || accountId.isBlank()) {
            throw new IllegalArgumentException("Account ID must not be null or blank");
        }
        if (transactions == null) {
            throw new IllegalArgumentException("Transactions list must not be null");
        }
        if (referenceTime == null) {
            throw new IllegalArgumentException("Reference time must not be null");
        }

        FraudSignalAssessment velocity = signalCalculator.calculateVelocity(transactions, referenceTime);
        FraudSignalAssessment geo = signalCalculator.calculateGeo(transactions, referenceTime);
        FraudSignalAssessment mcc = signalCalculator.calculateMcc(transactions);
        FraudSignalAssessment highValue = signalCalculator.calculateHighValue(transactions);

        double totalRiskScore = Math.min(
                velocity.weightedContribution()
                        + geo.weightedContribution()
                        + mcc.weightedContribution()
                        + highValue.weightedContribution(),
                1.0);

        List<FraudSignalAssessment> orderedSignals = List.of(velocity, geo, mcc, highValue).stream()
                .sorted(Comparator.comparingDouble(FraudSignalAssessment::weightedContribution).reversed())
                .toList();

        RiskTier riskTier = determineTier(totalRiskScore);
        String recommendedAction = determineRecommendedAction(riskTier);

        return new RiskAssessment(
                accountId,
                referenceTime.toString(),
                totalRiskScore,
                riskTier,
                orderedSignals,
                recommendedAction
        );
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private RiskTier determineTier(double totalRiskScore) {
        if (totalRiskScore >= config.getRedTierThreshold()) {
            return RiskTier.RED;
        }
        if (totalRiskScore >= config.getYellowTierThreshold()) {
            return RiskTier.YELLOW;
        }
        return RiskTier.GREEN;
    }

    private String determineRecommendedAction(RiskTier tier) {
        return switch (tier) {
            case RED -> "Block and escalate";
            case YELLOW -> "Review";
            case GREEN -> "Monitor";
        };
    }
}

File: frontend/src/app/core/services/risk-scoring.service.ts
import { Injectable, inject } from '@angular/core';
import { Transaction } from '../../shared/models/transaction.model';
import { FraudSignalAssessment, FraudSignalType } from '../../shared/models/fraud-signal-assessment.model';
import { RiskAssessment, RiskTier } from '../../shared/models/risk-assessment.model';
import { DataLoaderService } from './data-loader.service';
import { RISK_SCORING_CONFIG } from '../../shared/config/risk-scoring.config';

/**
 * Computes a deterministic RiskAssessment for a given account by evaluating
 * the four configured fraud signals against the account's recent transactions.
 */
@Injectable({ providedIn: 'root' })
export class RiskScoringService {
  private readonly dataLoader = inject(DataLoaderService);
  private readonly config = RISK_SCORING_CONFIG;

  getLatestRiskAssessment(accountId: string): RiskAssessment {
    const transactions = this.dataLoader.getTransactionsByAccountId(accountId);
    const now = this.referenceTime(transactions);
    const recent = this.last24Hours(transactions, now);

    const signals: FraudSignalAssessment[] = [
      this.velocitySignal(recent, now),
      this.geoSignal(recent),
      this.mccSignal(recent),
      this.highValueSignal(recent)
    ];

    const totalRiskScore = signals.reduce((sum, s) => sum + s.weightedContribution, 0);
    const clamped = Math.min(1.0, Math.max(0.0, Number(totalRiskScore.toFixed(4))));
    const riskTier = this.toTier(clamped);

    return {
      accountId,
      assessmentTimestamp: now.toISOString(),
      totalRiskScore: clamped,
      riskTier,
      contributingSignals: [...signals].sort(
        (a, b) => b.weightedContribution - a.weightedContribution
      ),
      recommendedAction: this.toRecommendedAction(riskTier)
    };
  }

  // ── Signal calculators ───────────────────────────────────────────────────

  private velocitySignal(transactions: Transaction[], now: Date): FraudSignalAssessment {
    const oneHourAgo = new Date(now.getTime() - 60 * 60 * 1000);
    const count = transactions.filter((t) => new Date(t.timestamp) >= oneHourAgo).length;
    const normalized = Math.min(1.0, count / this.config.velocityThreshold);
    const weight = this.config.weights.VELOCITY;
    return {
      signalType: 'VELOCITY' as FraudSignalType,
      rawValue: count,
      normalizedScore: round4(normalized),
      weight,
      weightedContribution: round4(normalized * weight),
      explanation: `${count} transaction${count !== 1 ? 's' : ''} in the last hour (threshold: ${this.config.velocityThreshold})`
    };
  }

  private geoSignal(transactions: Transaction[]): FraudSignalAssessment {
    const countries = new Set(
      transactions.map((t) => t.merchantCountry).filter((c): c is string => !!c)
    );
    const count = countries.size;
    const normalized = Math.min(1.0, count / this.config.geoThreshold);
    const weight = this.config.weights.GEO;
    return {
      signalType: 'GEO' as FraudSignalType,
      rawValue: count,
      normalizedScore: round4(normalized),
      weight,
      weightedContribution: round4(normalized * weight),
      explanation: `Transactions across ${count} distinct countr${count !== 1 ? 'ies' : 'y'} in 24 hours`
    };
  }

  private mccSignal(transactions: Transaction[]): FraudSignalAssessment {
    const highRisk = transactions.filter((t) =>
      this.config.highRiskMerchantCategories.has(t.merchantCategory)
    );
    const ratio = transactions.length === 0 ? 0 : highRisk.length / transactions.length;
    const normalized = Math.min(1.0, ratio);
    const weight = this.config.weights.MCC;
    return {
      signalType: 'MCC' as FraudSignalType,
      rawValue: highRisk.length,
      normalizedScore: round4(normalized),
      weight,
      weightedContribution: round4(normalized * weight),
      explanation: `${highRisk.length} of ${transactions.length} transaction${transactions.length !== 1 ? 's' : ''} at elevated-risk merchant categories`
    };
  }

  private highValueSignal(transactions: Transaction[]): FraudSignalAssessment {
    const maxAmount = transactions.reduce((m, t) => Math.max(m, t.amount), 0);
    const normalized = Math.min(1.0, maxAmount / this.config.highValueThreshold);
    const weight = this.config.weights.HIGH_VALUE;
    return {
      signalType: 'HIGH_VALUE' as FraudSignalType,
      rawValue: maxAmount,
      normalizedScore: round4(normalized),
      weight,
      weightedContribution: round4(normalized * weight),
      explanation: `Peak transaction amount: ${maxAmount.toLocaleString(undefined, { maximumFractionDigits: 2 })} (threshold: ${this.config.highValueThreshold.toLocaleString()})`
    };
  }

  // ── Tier and action mappers ──────────────────────────────────────────────

  private toTier(score: number): RiskTier {
    if (score >= this.config.thresholds.RED_MIN) return 'RED';
    if (score >= this.config.thresholds.YELLOW_MIN) return 'YELLOW';
    return 'GREEN';
  }

  private toRecommendedAction(tier: RiskTier): string {
    switch (tier) {
      case 'GREEN':
        return 'Monitor — no immediate action required.';
      case 'YELLOW':
        return 'Review — verify recent activity and apply additional scrutiny.';
      case 'RED':
        return 'Block and escalate — halt transactions and refer for manual investigation.';
      default:
        return 'Monitor — no action required.';
    }
  }

  // ── Helpers ──────────────────────────────────────────────────────────────

  private referenceTime(transactions: Transaction[]): Date {
    if (transactions.length === 0) return new Date();
    const maxTs = Math.max(...transactions.map((t) => new Date(t.timestamp).getTime()));
    return new Date(maxTs);
  }

  private last24Hours(transactions: Transaction[], now: Date): Transaction[] {
    const cutoff = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    return transactions.filter((t) => new Date(t.timestamp) >= cutoff);
  }
}

function round4(n: number): number {
  return Math.round(n * 10_000) / 10_000;
}
